pkg/sfu/streamallocator/probe_controller.go

```diff
@@ -207,9 +216,75 @@ func (p *ProbeController) InitProbe(probeGoalDeltaBps int64, expectedBandwidthUs
 		ProbeMaxDuration,
 	)
 
+	p.pollProbe(p.probeClusterId)
+
 	return p.probeClusterId, p.probeGoalBps
 }
 
+func (p *ProbeController) pollProbe(probeClusterId ProbeClusterId) {
+	p.lock.RLock()
+	if p.sendSideBWE == nil {
+		p.lock.RUnlock()
+		return
+	}
+
+	startingEstimate := p.sendSideBWE.GetEstimatedChannelCapacity()
+	p.lock.RUnlock()
+
+	go func() {
+		for {
+			p.lock.Lock()
+			if p.probeClusterId != probeClusterId {
+				p.lock.Unlock()
+				return
+			}
+
+			done := false
+			congestionState := p.sendSideBWE.GetCongestionState()
+			currentEstimate := p.sendSideBWE.GetEstimatedChannelCapacity()
+			switch {
+			case currentEstimate <= startingEstimate && time.Since(p.lastProbeStartTime) > ProbeTrendWait:
+				//
+				// More of a safety net.
+				// In rare cases, the estimate gets stuck. Prevent from probe running amok
+				// STREAM-ALLOCATOR-TODO: Need more testing here to ensure that probe does not cause a lot of damage
+				//
+				p.params.Logger.Infow("stream allocator: probe: aborting, no trend", "cluster", probeClusterId)
+				p.abortProbeLocked()
+				done = true
+				break
+
+			case congestionState == sendsidebwe.CongestionStateCongested || congestionState == sendsidebwe.CongestionStateEarlyWarning:
+				// stop immediately if the probe is congesting channel more
+				p.params.Logger.Infow("stream allocator: probe: aborting, channel is congesting", "cluster", probeClusterId)
+				p.abortProbeLocked()
+				done = true
+				break
+
+			case currentEstimate > p.probeGoalBps:
+				// reached goal, stop probing
+				p.params.Logger.Infow(
+					"stream allocator: probe: stopping, goal reached",
+					"cluster", probeClusterId,
+					"goal", p.probeGoalBps,
+					"current", currentEstimate,
+				)
+				p.StopProbe()
+				done = true
+				break
+			}
+			p.lock.Unlock()
+
+			if done {
+				return
+			}
+
+			// SSBWE-TODO: do not hard code sleep time
+			time.Sleep(50 * time.Millisecond)
+		}
+	}()
+}
```

Comment:
This is what triggered splitting out the probe controller from streamallocator. With REMB, we were checking status of probe when a new estimate is received from the remote side. With Pion TWCC also, that scheme worked as there is a bitrate estimate change callback.

But, with native TWCC, thinking of doing callbacks on congestion state changes only. It is possible that probes are handled separately and send side BWE module could fire callbacks on probe being successful/failure. If that happens, this code will not be necessary. But, an intermediate step here is to run a go routine to check on the status of the probe. Even if callbacks are used, spitting out the probe controller is a good thing as the code is easier to read (hopefully)